spec/index.bs

```diff
@@ -15686,29 +15686,49 @@ computes the fragment data (often a color) to be written into render targets.
 This stage produces a <dfn dfn>Fragment</dfn> for each [=RasterizationPoint=]:
 <div algorithm="Fragment accessors" dfn-for=Fragment>
     - <dfn dfn>destination</dfn> refers to [=FragmentDestination=].
+    - <dfn dfn>frontFacing</dfn> is true if it's a fragment on the front face of a primitive.
     - <dfn dfn>coverageMask</dfn> refers to multisample coverage mask (see [[#sample-masking]]).
     - <dfn dfn>depth</dfn> refers to the depth in [=viewport coordinates=],
         i.e. between the {{RenderState/[[viewport]]}} `minDepth` and `maxDepth`.
     - <dfn dfn>colors</dfn> refers to the list of color values,
         one for each target in {{GPURenderPassDescriptor/colorAttachments}}.
+    - <dfn dfn>depthPassed</dfn>
+        is `true` if the fragment passed the {{GPUDepthStencilState/depthCompare}} operation.
+    - <dfn dfn>stencilPassed</dfn>
+        is `true` if  the fragment passed the stencil {{GPUStencilFaceState/compare}} operation.
 </div>
 
 <div algorithm data-timeline=queue>
-    <dfn abstract-op>process fragment</dfn>(rp, desc, state)
+    <dfn abstract-op>process fragment</dfn>(rp, descriptor, state)
 
     **Arguments:**
 
     - |rp|: The [=RasterizationPoint=], produced by [[#rasterization]].
-    - |desc|: The descriptor of type {{GPUFragmentState}}.
+    - |descriptor|: The descriptor of type {{GPURenderPipelineDescriptor}}.
     - |state|: The active [=RenderState=].
 
     **Returns:** [=Fragment=] or `null`.
 
+    1. Let |fragmentDesc| be |descriptor|.{{GPURenderPipelineDescriptor/fragment}}.
+    1. Let |depthStencilDesc| be |descriptor|.{{GPURenderPipelineDescriptor/depthStencil}}.
     1. Let |fragment| be a new [=Fragment=] object.
     1. Set |fragment|.[=Fragment/destination=] to |rp|.[=RasterizationPoint/destination=].
+    1. Set |fragment|.[=Fragment/frontFacing=] to |rp|.[=RasterizationPoint/frontFacing=].
     1. Set |fragment|.[=Fragment/coverageMask=] to |rp|.[=RasterizationPoint/coverageMask=].
     1. Set |fragment|.[=Fragment/depth=] to |rp|.[=RasterizationPoint/depth=].
-    1. If |desc| is not `null`:
+    1. If `frag_depth` [=builtin=] is not produced by the shader:
+        1. Set |fragment|.[=Fragment/depthPassed=] to the result of [$compare fragment$](|fragment|.[=Fragment/destination=],
+            |fragment|.[=Fragment/depth=], depth, |state|.{{RenderState/[[depthStencilAttachment]]}},
+            |depthStencilDesc|?.{{GPUDepthStencilState/depthCompare}}).
+    1. Set |stencilState| to |depthStencilDesc|?.{{GPUDepthStencilState/stencilFront}} if
+        |rp|.[=RasterizationPoint/frontFacing=] is `true` and |depthStencilDesc|?.{{GPUDepthStencilState/stencilBack}}
+        otherwise.
+    1. Set |fragment|.[=Fragment/stencilPassed=] to the result of [$compare fragment$](|fragment|.[=Fragment/destination=],
```

Comment:
Do we know if early depth / stencil is forced to happen and will definitely skip the fragment (as an SSBO writes would not be visible)?

Reply:
Hm... you're right on the stencil buffer. From what I'm finding stencil tests CAN cause the fragment shader to be skipped, but won't if certain criteria is met (such as `discard` begin used.) I haven't found any documentation regarding whether or not SSBO write have the same effect.

Updated this line to only consider depth/`frag_depth` for the time being, since that's much more of a sure thing.